### ISSUE TITLE
[explicit-module-boundary-types] false positive error on array of functions in an object

### ISSUE THREAD

**Issue by danhorvath** [typescript-eslint/typescript-eslint#2701]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/explicit-module-boundary-types": [2]
  }
}
```

```TS
type Prop = {
  function: (input: number) => string
  functionArray: ((input: number) => string)[]
}

export const object: Prop = {
  function: (input) => String(input),
  functionArray: [(input) => String(input)],
}
```

**Expected Result**

No linter errors since everything is typed.

**Actual Result**

The linter shows errors for the anonymous function in `functionArray`:
`Missing return type on function.eslint@typescript-eslint/explicit-module-boundary-types`
`Argument 'input' should be typed.eslint@typescript-eslint/explicit-module-boundary-types`

The `function` field has no errors - I expect the same for the `functionArray` field.

**Additional Info**
N/A

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.4.1` |
| `@typescript-eslint/parser`        | `4.4.1` |
| `TypeScript`                       | `4.0.3` |
| `ESLint`                           | `7.11.0` |
| `node`                             | `14.13.1` |

**Comment by bradzacher** [comment]
Duplicate of #2183
